packages/contracts/evm-contracts/contracts/orderbook/OrderbookDex.sol

```diff
@@ -155,18 +222,48 @@ contract OrderbookDex is IOrderbookDex, ERC1155Holder, ReentrancyGuard {
             if (assetsToBuy == 0) {
                 continue;
             }
-            order.assetAmount -= assetsToBuy;
-            remainingEth -= assetsToBuy * order.pricePerAsset;
-            remainingAsset -= assetsToBuy;
+
+            // Can be unchecked because assetsToBuy is less than or equal to order.assetAmount.
+            unchecked {
+                order.assetAmount -= assetsToBuy;
+            }
+
+            uint256 purchaseCost = assetsToBuy * order.pricePerAsset;
+            uint256 makerFee = (purchaseCost * order.makerFee) / basisPoints;
+            uint256 takerFee = (purchaseCost * order.takerFee) / basisPoints;
+
+            if (remainingEth < purchaseCost + takerFee) {
+                revert InsufficientPayment();
+            }
+
+            // Can be unchecked because (purchaseCost + takerFee) is less than or equal to remainingEth.
+            unchecked {
+                remainingEth -= (purchaseCost + takerFee);
+            }
+            // Can be unchecked because assetsToBuy is less than or equal to remainingAsset.
+            unchecked {
+                remainingAsset -= assetsToBuy;
+            }
+
             IInverseProjected1155(asset).safeTransferFrom(
                 address(this),
                 msg.sender,
                 order.assetId,
                 assetsToBuy,
                 bytes("")
             );
-            order.seller.sendValue(assetsToBuy * order.pricePerAsset);
-            emit OrderFilled(order.seller, orderId, msg.sender, assetsToBuy, order.pricePerAsset);
+            order.seller.sendValue(purchaseCost - makerFee);
+            emit OrderFilled(
+                asset,
+                order.assetId,
+                orderId,
+                order.seller,
+                msg.sender,
+                assetsToBuy,
+                order.pricePerAsset,
+                makerFee,
+                takerFee
+            );
             if (remainingAsset == 0) {
```

Comment:
Same comment

Reply:
Same answer, although asset does not have decimals so the situation is less probable. Also floating point precision shouldn't make any problems if frontend properly operates with BigInt